### PR TITLE
[SYCL] Implement backend_traits (SYCL2020 interop pt2)

### PR DESCRIPTION
**Pull request by alexbatashev**
Link to the spec: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#_type_traits_backend_traits

### CODE REVIEW

**Comment by smaslov-intel** [comment]
@alexbatashev , please give a link to the spec this is implementing,

**Comment by alexbatashev** [comment]
> @alexbatashev , please give a link to the spec this is implementing,

Put it in the PR description

File: sycl/include/CL/sycl.hpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include <CL/sycl/aspects.hpp>
 #include <CL/sycl/atomic.hpp>
 #include <CL/sycl/backend.hpp>
+#include <CL/sycl/backend_traits.hpp>
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl.hpp:22]: Why did you choose to add this new header?

File: sycl/test/on-device/basic_tests/interop/traits.cpp [hunk 2]
```diff
@@ -0,0 +1,59 @@
+// RUN: %clangxx -fsycl -DUSE_OPENCL %s
```
Comment (smaslov-intel) [inline sycl/test/on-device/basic_tests/interop/traits.cpp:1]: What causes the force of the backend at run-time?

File: sycl/include/CL/sycl/backend.hpp [hunk 3]
```diff
@@ -63,15 +64,18 @@ template <backend Backend>
 typename std::enable_if<
     detail::InteropFeatureSupportMap<Backend>::MakePlatform == true,
     platform>::type
-make_platform(const typename interop<Backend, platform>::type &BackendObject) {
+make_platform(
+    const typename backend_traits<Backend>::template input_type<platform>
```
Comment (smaslov-intel) [inline sycl/include/CL/sycl/backend.hpp:68]: shouldn't we replace/remove the `interop` ?

File: sycl/test/on-device/basic_tests/interop/traits.cpp (hunk 2 above)
Comment (alexbatashev) [reply sycl/test/on-device/basic_tests/interop/traits.cpp:1]: It's a compile-only test, the program is not really run

File: sycl/include/CL/sycl.hpp (hunk 1 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl.hpp:22]: Would you like me to move the implementation to backend.hpp or just remove this particular line?

File: sycl/include/CL/sycl/backend.hpp (hunk 3 above)
Comment (alexbatashev) [reply sycl/include/CL/sycl/backend.hpp:68]: I'm not sure removing is a good idea. This seems to be publicly available code, that may have some users. Replacing all instances of `interop` makes sense to me, but I'd prefer to do it as a separate patch, as the change would be quite big.

File: sycl/include/CL/sycl.hpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl.hpp:22]: Yes, I meant that maybe we'd add this to backend.hpp?

File: sycl/include/CL/sycl/backend.hpp (hunk 3 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl/backend.hpp:68]: You are right, we can't remove this easily without breaking existing code.

File: sycl/include/CL/sycl.hpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/include/CL/sycl.hpp:22]: This is going to be a small class that never changes, so it doesn't make sense to keep it in its own file I think

**Review by smaslov-intel — APPROVED** [review]
Please consider moving backend_traits.hpp into backend.hpp

**Comment by alexbatashev** [comment]
> Please consider moving backend_traits.hpp into backend.hpp

Done

**Review by smaslov-intel — APPROVED** [review]

**Pull request merged**